// @shopify/app/dist/commands/app/build.js
import { a as appFlags, l as load } from '../../app-e1e50846.js';
import { b as buildExtension } from '../../extension-54b7397f.js';
import { system, output, cli, path } from '@shopify/cli-kit';
import { i as installAppDependencies } from '../../dependencies-8386bbeb.js';
import { Flags, Command } from '@oclif/core';
import '../../cli-852b5950.js';
import 'node:zlib';
import 'node:fs';
import 'node:stream';
import 'node:util';
import 'url';
import '../../configuration-a702715b.js';

async function web(command, { web: web2, stdout, stderr, signal }) {
  const script = web2.configuration.commands[command];
  if (!script) {
    return;
  }
  const [cmd, ...args] = script.split(" ");
  await system.exec(cmd, args, { cwd: web2.directory, stdout, stderr, signal });
  stdout.write("Web successfully built.");
}

async function build({ app, skipDependenciesInstallation }) {
  if (!skipDependenciesInstallation) {
    await installAppDependencies(app);
  }
  await output.concurrent([
    ...app.webs.map((web$1) => {
      return {
        prefix: web$1.configuration.type,
        action: async (stdout, stderr, signal) => {
          await web("build", { web: web$1, stdout, stderr, signal });
        }
      };
    }),
    {
      prefix: "extensions",
      action: async (stdout, stderr, signal) => {
        await buildExtension({ app, extensions: app.extensions.ui, stdout, stderr, signal });
      }
    }
  ]);
  output.newline();
  output.success(`${app.name} built`);
}

const _Build = class extends Command {
  async run() {
    const { flags } = await this.parse(_Build);
    const directory = flags.path ? path.resolve(flags.path) : process.cwd();
    const app = await load(directory);
    await build({ app, skipDependenciesInstallation: flags["skip-dependencies-installation"] });
  }
};
let Build = _Build;
Build.description = "Build the app";
Build.flags = {
  ...cli.globalFlags,
  ...appFlags,
  "skip-dependencies-installation": Flags.boolean({
    hidden: false,
    description: "Skips the installation of dependencies.",
    env: "SHOPIFY_FLAG_SKIP_DEPENDENCIES_INSTALLATION",
    default: false
  })
};

export { Build as default };
